File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        n = len(heights) - 1
        l, r = 0, n
        max_area = -1

        while l <= r :
            width = r - l
            height = min(heights[r], heights[l])
            curr_max_area = width * height
            max_area = max(max_area, curr_max_area)
            if heights[l] > heights[r]:
                r -= 1
            else:
                l += 1

        return max_area
               



        

File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n {static_cast<int>(height.size())};

        vector<int> left_max(n, 0);
        vector<int> right_max(n, 0);

        left_max[0] = height[0];
        for (int i{1}; i < n; ++i) {
            left_max[i] = std::max(left_max[i-1], height[i]);
        }

        right_max[n-1] = height[n-1];
        for (int i{n-2}; i >= 0; --i) {
            right_max[i] = std::max(right_max[i+1], height[i]);
        }
        int res{0};
        for (int i{0}; i < n; i++) {
            res +=  std::min(left_max[i], right_max[i]) - height[i];
        }

        return res;
    }
};
